### PR TITLE
Update network list logic and sorting

### PR DESCRIPTION
**Pull request by zerts**

### CODE REVIEW

File: src/modules/ethereum/chains/types.ts [hunk 1]
```diff
@@ -17,9 +17,17 @@ export interface ChainConfigV1 {
...
   /** @deprecated */
   ethereumChains?: DeprecatedEthereumChainConfig[];
 }
+
+export interface ChainConfig {
+  version: 3;
+  ethereumChainConfigs: EthereumChainConfig[];
+  visitedChains: string[];
+  /** @deprecated */
+  ethereumChains?: DeprecatedEthereumChainConfig[];
```
Comment (zerts) [inline src/modules/ethereum/chains/types.ts:32]: Should we exclude this field from the type?

File: src/modules/ethereum/chains/types.ts [hunk 2]
```diff
@@ -24,7 +24,12 @@ export interface ChainConfigV2 {
...
 
-export interface ChainConfig extends Omit<ChainConfigV2, 'version'> {
+export interface ChainConfigV3 extends Omit<ChainConfigV2, 'version'> {
   version: 3;
   /** This version fixes the bug with undefined ids for old custom networks */
 }
+
+export interface ChainConfig extends Omit<ChainConfigV3, 'version'> {
+  version: 4;
+  visitedChains: string[];
```
Comment (zerts) [inline src/modules/ethereum/chains/types.ts:34]: Make optional

Comment (zerts) [reply src/modules/ethereum/chains/types.ts:34]: To remove migration

File: src/modules/networks/networks-store.ts [hunk 3]
```diff
@@ -29,30 +29,31 @@ export class NetworksStore extends Store<State> {
   private networkConfigs: NetworkConfig[] = [];
   private customNetworkConfigs: NetworkConfig[] = [];
   private loaderPromises: Record<string, Promise<Networks>> = {};
-  private getEthereumChainConfigs:
-    | null
-    | (() => Promise<EthereumChainConfig[] | undefined>);
+  private getChainSources: null | (() => Promise<ChainConfig | undefined>);
```
Comment (zerts) [inline src/modules/networks/networks-store.ts:32]: Declare type more explicit

File: src/modules/networks/networks-store.ts [hunk 4]
```diff
@@ -74,11 +75,19 @@ export class NetworksStore extends Store<State> {
...
-    const savedChainConfigs = await this.getEthereumChainConfigs?.();
+    const chainConfigs = await this.getChainSources?.();
+    const savedChainConfigs = chainConfigs?.ethereumChainConfigs;
+    const visitedChains = chainConfigs?.visitedChains;
     const savedIds = savedChainConfigs?.map((config) => config.id);
     const chainsToFetch = Array.from(
-      new Set([...(savedIds || []), ...(chains || [])])
-    ).filter((id) => !isCustomNetworkId(id));
+      new Set(
+        [
```
Comment (zerts) [inline src/modules/networks/networks-store.ts:84]: Can this be made more beautiful?

File: src/ui/components/NetworkSelectDialog/createNetworkGroups.ts [hunk 5]
```diff
@@ -45,17 +43,12 @@ export function createGroups({
   filterPredicate?: (network: NetworkConfig) => boolean;
   sortMainNetworksType?: 'alphabetical' | 'by_distribution';
 }): NetworkGroups {
-  const mainnetList = networks.getMainnets().filter(filterPredicate);
-  const allNetworks = networks.getNetworks().filter(filterPredicate);
+  const mainnetList = networks.getUserNetworks().filter(filterPredicate);
+  const allNetworks = networks.getUserNetworks().filter(filterPredicate);
```
Comment (zerts) [inline src/ui/components/NetworkSelectDialog/createNetworkGroups.ts:47]: Why there are 2 same lines?

File: src/ui/components/NetworkSelectDialog/createNetworkGroups.ts [hunk 6]
```diff
@@ -45,17 +43,12 @@ export function createGroups({
...
   const testnetList = networks.getTestNetworks().filter(filterPredicate);
   const mainNetworkPredicate = (network: NetworkConfig) => {
```
Comment (zerts) [inline src/ui/components/NetworkSelectDialog/createNetworkGroups.ts:49]: Rename?

File: src/modules/networks/networks-store.ts [hunk 7]
```diff
@@ -30,26 +30,24 @@ export class NetworksStore extends Store<State> {
...
   constructor(
     state: State,
     {
-      getEthereumChainConfigs,
+      getChainSources,
       client,
       testnetMode,
     }: {
-      getEthereumChainConfigs?: NetworksStore['getEthereumChainConfigs'];
+      getChainSources?: NetworksStore['getChainSources'];
```
Comment (everdimension) [inline src/modules/networks/networks-store.ts:44]: I was thinking about a better name and so far have come up with `"otherNetworksData"`
So this method would accordingly be named `getOtherNetworksData`

What do you think?

Comment (zerts) [reply src/modules/networks/networks-store.ts:44]: I've though more about `getSavedNetworkData`.
Cause what we get here is everything we've saved about networks configs/visits/etc

File: src/modules/networks/networks-store.ts [hunk 8]
```diff
@@ -26,30 +26,35 @@ function mergeNetworkConfigs(
   ];
 }
 
+type ChainConfigSource = {
```
Comment (everdimension) [inline src/modules/networks/networks-store.ts:29]: "SavedNetworksData" or "OtherNetworksData"?

File: src/modules/networks/networks-store.ts [hunk 9]
```diff
@@ -26,30 +26,35 @@ function mergeNetworkConfigs(
...
+  ethereumChainConfigs: EthereumChainConfig[];
+  visitedChains?: string[];
```
Comment (everdimension) [inline src/modules/networks/networks-store.ts:31]: maybe it's better to have `visitedChains: string[] | null` here?

**Review by everdimension — APPROVED** [review]

**Review by everdimension — APPROVED** [review]

**Pull request merged**